neqo-http3-server/src/main.rs

```diff
@@ -207,234 +209,299 @@ fn process(
+            args,
+            poll: Poll::new()?,
+            sockets: Vec::new(),
+            servers: HashMap::new(),
+            timer: Builder::default()
+                .tick_duration(Duration::from_millis(1))
+                .build::<usize>(),
+            hosts: Vec::new(),
+            active_servers: HashSet::new(),
+        })
+    }
 
-    let buf = &mut [0u8; 2048];
+    /// Init Poll for all hosts. Create sockets, and a map of the
+    /// socketaddrs to instances of the HttpServer handling that addr.
+    pub fn init(&mut self) -> Result<(), io::Error> {
+        self.hosts = self.args.host_socket_addrs();
+        if self.hosts.is_empty() {
+            eprintln!("No valid hosts defined");
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "No hosts"));
+        }
 
-    let mut events = Events::with_capacity(1024);
-
-    loop {
-        poll.poll(&mut events, None)?;
-        for event in &events {
-            if event.token() == TIMER_TOKEN {
-                while let Some(inx) = timer.poll() {
-                    if let Some(socket) = sockets.get_mut(inx) {
-                        qinfo!("Timer expired for {:?}", socket);
-                        if let Some((ref mut server, svr_timeout)) =
-                            servers.get_mut(&socket.local_addr().unwrap())
-                        {
-                            process(&mut **server, svr_timeout, inx, None, &mut timer, socket);
-                        }
-                    }
+        for (i, host) in self.hosts.iter().enumerate() {
+            let socket = match UdpSocket::bind(&host) {
+                Err(err) => {
+                    eprintln!("Unable to bind UDP socket: {}", err);
+                    return Err(err);
                 }
-            } else if let Some(socket) = sockets.get_mut(event.token().0) {
-                let local_addr = hosts[event.token().0];
+                Ok(s) => s,
+            };
 
-                if !event.readiness().is_readable() {
-                    continue;
+            let local_addr = match socket.local_addr() {
+                Err(err) => {
+                    eprintln!("Socket local address not bound: {}", err);
+                    return Err(err);
                 }
+                Ok(s) => s,
+            };
+
+            let res = socket.only_v6();
+            let also_v4 = if res.is_ok() && !res.unwrap() {
+                " as well as V4"
+            } else {
+                ""
+            };
+            println!(
+                "Server waiting for connection on: {:?}{}",
+                local_addr, also_v4
+            );
+
+            self.poll.register(
+                &socket,
+                Token(i),
+                Ready::readable() | Ready::writable(),
+                PollOpt::edge(),
+            )?;
+
+            self.sockets.push(socket);
+            self.servers
+                .insert(local_addr, (create_server(&self.args), None));
```

Comment:
create_server can be an associated function, rather than a completely separate function.  That is `Self::create_server` or `self.create_server`.